R/epiparameter_db.R

```diff
@@ -0,0 +1,765 @@
+    return(
+      list(
+        params = NA_real_,
+        uncertainty = list(uncertainty = create_uncertainty())
+      )
+    )
+  }
+
+  # return params without uncertainty
+  if (!any(grepl(pattern = "ci", x = names(params), fixed = TRUE))) {
+    uncertainty <- lapply(
+      vector(mode = "list", length = length(params)),
+      function(x) x <- create_uncertainty()
+    )
+    names(uncertainty) <- names(params)
+    return(
+      list(
+        params = unlist(params),
+        uncertainty = uncertainty
+      )
+    )
+  }
+
+  # add ci to params without for consistent data struct
+  params_ <- vector(mode = "list")
+  for (i in seq_along(params)) {
+    nm <- names(params)[i]
+    nms <- names(params)[grepl(pattern = nm, x = names(params))]
+    if (any(grepl(pattern = "_ci", x = nms, fixed = TRUE))) {
+      params_ <- append(params_, params[i])
+    } else {
+      params_ci <- list(params[[i]], c(NA_real_, NA_real_), NA_real_)
+      names(params_ci) <- c(nm, paste0(nm, "_ci_limits"), paste0(nm, "_ci"))
+      params_ <- append(params_, params_ci)
+    }
+  }
+  params <- params_
+
+  # params and uncertainty
+  ci_limits <- params[!is.na(names(params)) &
+    endsWith(names(params), suffix = "ci_limits")]
+  ci <- params[!is.na(names(params)) & endsWith(names(params), suffix = "ci")]
+  ci_type <- switch(x$metadata$inference_method,
+    mle = "confidence_interval",
+    bayesian = "credible_interval",
+    NA_character_
+  )
+
+  # separate parameters and uncertainty
+  ci_idx <- grepl(pattern = "_ci", x = names(params), fixed = TRUE)
+  uncertainty <- params[ci_idx]
+  params <- params[!ci_idx]
+
+  # default uncertainty for each parameters
+  uncertainty <- lapply(
+    vector(mode = "list", length = length(params)),
+    function(x) x <- create_uncertainty()
+  )
+
+  for (i in seq_along(ci_limits)) {
+    uncertainty[[i]] <- list(
+      ci_limits = ci_limits[[i]],
+      ci = ci[[i]],
+      ci_type = ci_type
+    )
+  }
+
+  params <- unlist(params)
+  names(uncertainty) <- names(params)
+
+  # return params and uncertainty
+  list(
+    params = params,
+    uncertainty = uncertainty
+  )
+}
+
+#' Check `<epiparameter>` based on condition
+#'
+#' @return list of logical booleans.
+#' @keywords internal
+#' @noRd
+.is_cond_epiparameter <- function(lst, condition, nse_subject) {
+  # check input
+  stopifnot(is_epiparameter(lst))
+  if (nse_subject == "prob_dist") {
+    # special case to subset by dist as name is extracted first
+    if (is.object(lst$prob_distribution) ||
+        is.character(lst$prob_distribution)) {
+      prob_dist <- family(lst) # nolint prob_dist used in eval
+      eval(expr = condition)
+    } else {
+      FALSE
+    }
+  } else {
+    # apply condition to <epiparameter> name matching elements
+    lapply(lst, function(x) {
+      if (nse_subject %in% names(x)) {
+        # <epiparameter> is only nested once so no need for recursive search
+        eval(expr = condition, envir = x)
```

Comment:
Feels slightly vulnerable to evaluate an expression without adding some checks to it. It is an opportunity for some code to be injected unknowingly. If you can add a check to ensure it's (semi) in the format you would expect, that would make the function more robust as well.

Reply:
I like this suggestion, but nothing comes immediately to mind as to how we could test this conditions before they get evaluated. Do you have ideas for how to do this?